ehsm_kms_service/access_control_server.js

```diff
@@ -0,0 +1,194 @@
+const logger = require("./logger")
+
+const {
+  EHSM_CONFIG_DENY_SIZE_LIMIT,
+  EHSM_CONFIG_DENY_TIMEOUT,
+  EHSM_CONFIG_DENY_COUNT_LIMIT,
+  EHSM_CONFIG_REPORTERS
+} = process.env
+
+const defense_apis = {
+  CheckAccess: 'CheckAccess',
+  ReportTarget: 'ReportTarget',
+  CheckNonce: 'CheckNonce'
+}
+
+const default_config_value = {
+  size_limit: 3000,
+  time_out: 1800000, //ms defult 30 min
+  count_limit: 5
+}
+
+const _DenyList = new Map()
+const _Reporters = () => {
+  if (EHSM_CONFIG_REPORTERS && EHSM_CONFIG_REPORTERS.length > 0) {
+    return EHSM_CONFIG_REPORTERS.split('|')
+  }
+  return []
+}
+const _result = (code, msg, data = {}) => {
+  return {
+    code: code,
+    message: msg,
+    result: {
+      ...data,
+    },
+  }
+}
+
+function DenyListClearUp() {
+  timestamp = new Date().getTime()
+  let _size_limit = parseInt(EHSM_CONFIG_DENY_SIZE_LIMIT)
+  if (isNaN(_size_limit)) {
+    _size_limit = default_config_value.size_limit
+  }
+  let _time_out = parseInt(EHSM_CONFIG_DENY_TIMEOUT)
+  if (isNaN(_time_out)) {
+    _time_out = default_config_value.time_out
+  }
+
+  let keys = Array.from(_DenyList.keys())
+  for (var i = 0; i < keys.length; i++) {
+    var key = keys[i]
+    if (_DenyList.has(key)) {
+      var report_info = _DenyList.get(key)
+      if (report_info.timestamp > 0 &&
+        timestamp > 0 &&
+        (report_info.timestamp + _time_out) < timestamp) {
+          _DenyList.delete(key)
+          keys.splice(i, 1)
+          i--
+        }
+    }
+  }
+
+  if (keys.length >= _size_limit) {
+    for (var i = 0; i < (keys.length - _size_limit); i++) {
+      if(_DenyList.has(keys[i])) {
+        _DenyList.delete(keys[i])
+      }
+    }
+  }
+}
+
+const DefenseCheckAccess = (ip) => {
+  DenyListClearUp()
+  let count_limit = parseInt(EHSM_CONFIG_DENY_COUNT_LIMIT)
+  if (isNaN(count_limit)) {
+    count_limit = default_config_value.count_limit
+  }
+  return (!_DenyList.has(ip) || (_DenyList.get(ip).count < count_limit))
+}
+
+const DefenseReportTarget = (ip, timestamp) => {
+  DenyListClearUp()
+  let count = 0
+  if (_DenyList.has(ip)) {
+    var info = _DenyList.get(ip)
+    count = info.count
+
+    if (count < 0) {
+      count = 0
+    }
+
+    if (info.timestamp < timestamp) {
+      _DenyList.delete(ip)
+    }
+  }
+
+  count = count + 1
+  if(!_DenyList.has(ip)) {
+    _DenyList.set(ip,
+      {
+        timestamp:timestamp,
+        count:count
+      })
+  }
+}
+const DefenseCheckNonce = (nonce_database, appid, nonce, nonce_timestamp) => {
+  const nonce_data = {
+    nonce: nonce,
+    nonce_timestamp: nonce_timestamp
+  }
+
+  if (!nonce_database[appid]) {
```

Comment:
not found the logic to delete the nonce after it's expired, e.g.
https://github.com/intel/ehsm/blob/main/ehsm_kms_service/function.js#L54

Reply:
Yes. We used _nonce_cache_timer() in function.js.
You can see it at ehsm_kms_server.js line 32 and ehsm_access_server.js line 16